python/ray/data/dataset.py

```diff
@@ -2842,6 +2851,82 @@ def _experimental_lazy(self) -> "Dataset[T]":
         self._lazy = True
         return self
 
+    def is_out_of_band_serializable(self) -> bool:
+        """Whether this dataset is able to be out-of-band serialized, i.e. serialized
+        for use across different Ray clusters. Only datasets that read from lazy
+        datasources (i.e. via one of the ray.data.read_*() APIs) are out-of-band
+        serializable.
+        """
+        return self._plan.has_lazy_input()
+
+    @DeveloperAPI
+    def serialize_out_of_band(self) -> bytes:
```

Comment:
Hmm I get what you're saying, although I think that these functions should be named for what the end-user is trying to do (using datasets "out-of-band", i.e. across clusters), without leaking the implementation detail that we're serializing the lineage. Ideally the user shouldn't care/have to know about that detail.

Agreed that the "out-of-band" bit is not super clear. What if these were renamed to `{de}serialize_for_cross_cluster`?

Reply:
How about using "export" as a verb here?

Reply:
I don't see "serialize lineage" is leaking internal details, "lineage" or "execution plan" as relatively well known concepts can be explained more easily than "out of band". I don't know any data processing framework can hide execution plan concept from users.

For serialize_for_cross_cluster, it can provoke questions like:
What is serialized exactly?  (user should not think the actual content of dataset gets serialized)
Does it mean it cannot be deserialized back in the same cluster? (I didn't see why it cannot -- if so, I think it's better not to dictate that in the method name)

Reply:
**TL;DR:** Although `serialize_for_storage()` gives us a more evolvable API with implementation flexibility and hints exactly what the serialization hooks are used for (persisting to storage), I think that `serialize_lineage()` is probably a better choice since the fact that this serialization mandates recomputation via lineage is a very important behavioral contract, and we don't want to surprise users with unexpected recomputation.

@ericl hmm for me, "export" connotes (1) persisting to a file, and (2) exporting for use outside of Ray, rather than serializing it to in-memory bytes for use across clusters, so I'm not sure if that verb is a good fit. 🤔 

>I don't see "serialize lineage" is leaking internal details, "lineage" or "execution plan" as relatively well known concepts can be explained more easily than "out of band". I don't know any data processing framework can hide execution plan concept from users.

@jianoaix don't all data processing frameworks do their best to hide their execution plan from users, when possible, e.g. for most APIs where the execution plan isn't relevant?

>What is serialized exactly? (user should not think the actual content of dataset gets serialized)

I suppose whether lineage/our execution model is an implementation detail depends upon whether we consider (1) the content of the serialized bytes, and (2) recomputation being required upon deserialization, to be part of the API contract. The intention is that: for (1), what is actually serialized is left as an implementation detail, where an opaque set of bytes is returned, since in the future, we might dynamically choose exactly _what_ is serialized based on dataset size, recomputation cost, etc.; for (2), whether recomputation occurs is also left as an implementation detail, open to transparent optimization by Datasets. More on both of these below.

>Does it mean it cannot be deserialized back in the same cluster? (I didn't see why it cannot -- if so, I think it's better not to dictate that in the method name)

It actually can, and often will be! Agreed on `for_cross_cluster` not being a good name.

**Zooming back out here so we all have the same context:** the core use case that we're targeting is saving a dataset to persistent storage for later use, i.e. checkpointing, and this will primarily be used by Tune when stopping and resuming experiments. We have two big constraints here:
- we can't save the actual data, since the data might be extremely large;
- we can't save the data futures since (1) our reference counting doesn't play nice with futures serialized directly with pickle (infinite pin), and (2) the dataset might be deserialized on another cluster which would render those futures invalid.

Therefore, we currently save the lineage required to recreate the dataset. But in the future, we might change this implementation: e.g. we might serialize the data directly if it's small, or transparently save the data to external storage (e.g. S3) if it's large and the recomputing cost is untenable, etc.

With this context, we have a few options:
1. `serialize_lineage()` - **name it for what it does and how its implemented**
2. `serialize_for_storage()` - **name it for what it does and what its used for**

(1) makes the current implementation part of the API contract, while (2) gives us some flexibility for changing our implementation in the future. However, the general size of the serialized bytes (e.g. a few hundred bytes for lineage vs 1 GB for the data), and lineage-based serialization mandating recomputation upon deserialization, are both very important parts of the API's behavioral contracts, and we don't want to surprise users with unexpected serialized sizes or recomputation on deserialization, **so I'm thinking that (1) would make the most sense for now.**

Eventually, I think that we'll want something like a `serialize_for_storage()` API that will try to choose the best implementation based on data size and recomputation cost, as alluded to above, with args exposed that can tune or override this behavior. But to start, `serialize_lineage()` seems like a better route.

Reply:
Thanks for thinking more about APIs.
For the verb, I don't have an opinion using import/export v.s. serialize/deserialize:)
The rest sound good to me; but I added one more option (see below) that we can consider.

I searched around a bit:
- For Polars that we are going to talk, on the Introduction and Getting Start pages (https://pola-rs.github.io/polars-book/user-guide/quickstart/intro.html), the two most prominent places for users, it turns out both have a mention of query plan. This effectively put the users in a position to know this concept (although details are not required) -- in terms of abstraction, this is the same as exposing an API with a mention of "query plan". I think they do not have such API yet, but that's largely a usecase thing.
- For more mature frame Spark, it already has an option exposed to user called spark.logLineage (for different use cases).

@clarkzinzow Another option I can think of is serialize_as/to_lazy_dataset / deserialize_from_lazy_dataset (or remove as/to/from if that makes more sense), because:
- "lazy dataset" is also a concept users will know (but maybe at slightly higher abstraction? actually to explain laziness, lineage/execution plan is used -- another reason I don't concern about letting users know lineage)
- it says clear that it's not dumping the whole lot of data
- if in the future we do want to dump the whole actual data, we can add a new pair of APIs for that

Reply:
All data processing frameworks will _mention_ their query/execution planner in their docs, since they want to emphasize optimized performance as one of their value-adds and the execution model may be relevant for debugging, setting performance expectations, etc. But that's different than requiring the user to think about their execution plan when using their APIs. The latter should only be done when required, for the sake of the typical tenets of good API design: information hiding, API should be as small/simple as possible, don't introduce unnecessary concepts, keep API concepts as high-level as possible, don't leak implementation details, etc.

This is all moot since the serialization being lineage-based seems to be a fundamental behavioral part of the API contract, so including the lineage concept in the serialization APIs seems like the best route.

>Another option I can think of is serialize_as/to_lazy_dataset / deserialize_from_lazy_dataset 

I'd push against that since this isn't correct in terms of what's supported, since lineage-based serialization is also supported for eager datasets. We track the lineage even when in eager mode, and when calling `ds.serialize_lineage()` for a dataset in eager mode, only the lineage will be serialized. The only requirement is that the dataset is **created** from external data, which we capture in the lineage as read tasks and is currently equivalent to lazy datasources.

Reply:
> All data processing frameworks will _mention_ their query/execution planner in their docs, since they want to emphasize optimized performance as one of their value-adds and the execution model may be relevant for debugging, setting performance expectations, etc. But that's different than requiring the user to think about their execution plan when using their APIs. The latter should only be done when required, for the sake of the typical tenets of good API design: information hiding, API should be as small/simple as possible, don't introduce unnecessary concepts, keep API concepts as high-level as possible, don't leak implementation details, etc.

I don't think just mentioning execution plan will mean performance or value added automatically -- it's just a standard thing across frameworks. Even SQL would have a statement like EXPLAIN  PLAN as part of the language for users.

The rest are just general principles of API design -- those are separate topics from whether lineage/execution plan is leaky.

> 
> This is all moot since the serialization being lineage-based seems to be a fundamental behavioral part of the API contract, so including the lineage concept in the serialization APIs seems like the best route.

Yes, exactly.

> 
> > Another option I can think of is serialize_as/to_lazy_dataset / deserialize_from_lazy_dataset
> 
> I'd push against that since this isn't correct in terms of what's supported, since lineage-based serialization is also supported for eager datasets. We track the lineage even when in eager mode, and when calling `ds.serialize_lineage()` for a dataset in eager mode, only the lineage will be serialized. The only requirement is that the dataset is **created** from external data, which we capture in the lineage as read tasks and is currently equivalent to lazy datasources.

That's why there is "as/to" in API name, so that it can apply to eager datasets as well.
That said, I'm fine with either choices, feel free to stick with serialize_lineage if you want.

Reply:
> I don't think just mentioning execution plan will mean performance or value added automatically -- it's just a standard thing across frameworks.

I was juxtaposing talking about a query/execution planner in high-level docs vs API-level docs, I know that just mentioning a planner won't automatically convey what I mentioned. But the query/execution planner is often discussed in landing-page/concepts/user guide docs as (1) a fundamental concept that can help the user to understand how their query is executed, and (2) a way to talk about their performance optimizations, among other reasons.

>Even SQL would have a statement like EXPLAIN PLAN as part of the language for users.

That's an API for explaining the execution plan, which would obviously deal with the execution plan. But other APIs that don't need to reference the execution plan, shouldn't for the sake of not needlessly confusing the user, right? I don't think I'm saying anything too controversial here.

>The rest are just general principles of API design -- those are separate topics from whether lineage/execution plan is leaky.

I think that each are relevant to this general API discussion, but we agree that including the lineage concept is not leaky, so we don't have to go into that.

>That's why there is "as/to" in API name, so that it can apply to eager datasets as well.

So I think where we're seeing this differently is that I don't consider lineage-based serialization to be converting the dataset to a lazy dataset. The dataset's execution mode will be maintained upon deserialization, e.g. if it's an eager dataset before lineage-based serialization, it will be eager after deserialization, so it's not really converting to or casting as a lazy dataset. I consider lineage as distinctly separate from execution mode.